projects/output-and-error-handling/README.md

```diff
@@ -0,0 +1,195 @@
+
+Another example is when something goes wrong - in your score-recording program, you may want to (reasonably) assume that anything it outputs is a number. But if something goes wrong (say, your network connection was down so the scores couldn't be fetched), reporting that on standard error means you won't accidentally try to add the error string "Network was down" to some other number. (TODO: This explanation doesn't seem very clear).
+
+#### Standard out and standard error in Go
+
+In go, standard out and standard error can be accessed as `os.Stdout` and `os.Stderr`.
+
+You can write to them by writing code like `fmt.Fprint(os.Stdout, "Hello")` or `fmt.Fprint(os.Stderr, "Something went wrong")`. (The "F" before "printf" stands for "file" - we're saying "print some string to a file I'll specify as the first argument". A lot of times in Unix systems, we like to pretend anything we read or write is a file).
+
+More often, we'll write `fmt.Print("Hello")` - this is the same as writing `fmt.Fprint(os.Stdout, "Hello")` (if you look at [the Go standard library source code](https://cs.opensource.google/go/go/+/refs/tags/go1.19.5:src/fmt/print.go;l=251-253), you can see it's literally the same), but it's worth remembering we can choose to write to other locations, like standard error, if it's more appropriate.
+
+#### When to write to standard out/error
+
+As a rule, the intended output of your program should be written to standard out, and anything that isn't the intended output of your program should be written to standard error.
+
+Some things you may write to standard error:
+* Progress messages explaining what the program is doing.
+* Error information about something that went wrong.
+
+Thinking about our program we're going to write, that means we're likely to write:
+* The current weather to standard out - it's what our program is for.
+* A message saying that we've been asked to wait and retry later to standard error - it's a progress message, not the intended output of our program.
+* Error information if the server seems broken to standard error - it's not the intended output of our program, it's diagnostic information.
+
+### Exit codes
+
+By convention, most programs exit with an exit code of `0` when they successfully did what they expected, and any number that isn't `0` when they didn't. Often a specific program will attach specific meaning to specific non-zero exit codes (e.g. `1` may mean "You didn't specify all the flags I needed" and `2` may mean "A remote server couldn't give me information I needed"), but there are no general conventions for specific non-`0` exit codes across different programs.
+
+By default, your program will exit with exit code `0` unless you tell it to do otherwise, or it crashes.
+
+In Go, you can choose what code to exit your program with by calling [`os.Exit`](https://pkg.go.dev/os#Exit). After calling `os.Exit`, your program stops and can't do anything else.
```

Comment:
worth noting how to check an exit code - usually by runnign echo $?
maybe suggest they run something from command line - pwd or ls or something that will succeed - and check it. 
then run something that fails - cd /doesnotexist and same to see failure

Reply:
Done - also included some example manpage sections showing how people document exit codes.